### PR TITLE
Topic recognition

### PR DESCRIPTION
**Pull request by jdw5**
## Project
This pull request provides a VQVAE implementation on the ADNI dataset (Task 7). In addition to the VQVAE, a GAN network is provided to act as a prior model to the VQVAE. The GAN network is trained to generate latent space representations of the training data, which are then used as input to the VQVAE network to generate images from provided dataset. 

## Structure
- `README.md` provides an overview of the project, the report, usage and complete documentation of all classes
- `dataset.py` contains class-based implementations for loading and preprocessing the dataset.
- `modules.py` contains class-based implementations of all networks used in the project.
- `train.py` contains the training loop and evaluation code.
- `predict.py` contains the code for generating images from both GAN and VQVAE model, as well as assessing the performance of the end-models
- `utils.py` contains a single source of truth for defining constants and parameters across all models
- `main.py` is the test driver for the codebase. It is used to run the training and evaluation code.
- `images/` is a top-level directory containing images for the results and the `README.md` report

## Results
The produced VQVAE was able to reconstruct the training set with no obvious differences to the original data when viewed manually. A GAN network was used to assist in generating images from noise by learning the probabilities of latent space representation as an input to the VQVAE network. This produced images of similar structure to the dataset, but with a noticeable blurring of portions and lack of fine details. The average SSIM was 0.62, as the structure of the image did resemble that of the test set data. 

Better results may be achievable with data augmentation to improve the representation of details in the encoded indices, and by changing the prior network used for generation as the generator diverged.

### CODE REVIEW

**Comment by SiyuLiu0329** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Hard

Readme Overall:  Good
- Project Overview: Good
- Model: Good
- Data + Preprocessing: Good
- Usage: Good
- Training / Loss Curve(s): Good
- Result Demonstration: Good
- References: Good

Functionality:
- Reconstruction: Good
- Generation: Good

Code:
- Consistent with Results: Yes
- File Structure: Good
- Commenting: Good
- Commit frequency: Good
- Commit messages: Good

Other Comments:
N/A

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
